src/meshTools/AMIInterpolation/AMIInterpolation/AMIMethod/directAMI/directAMI.C

```diff
@@ -24,6 +24,87 @@ License
 \*---------------------------------------------------------------------------*/
 
 #include "directAMI.H"
+#include "primitiveFieldsFwd.H"
+#include "pointFieldFwd.H"
+#include "faceData.H"
+#include "primitiveFieldsFwd.H"
+#include "faceFunctors.H"
+#include "gpuList.H"
+#include "scalarList.H"
+
+namespace Foam {
+    struct ShootRayFunctor {
+        const point *tgtPoints;
+        const point *srcPoints;
+        const vector *srcCf;
+        const faceData *tgtFaces;
+        const faceData *srcFaces;
+
+        // Arrays used for addressing
+        label* srcAddr;
+        label* tgtAddr;
+
+        // Arrays used to hold the actual count
+        label* srcCount;
+        label* tgtCount;
+
+        faceRayFunctor rayFunctor;
+        faceNormalFunctor normalFunctor;
+
+        const label srcSize;
+        const label tgtSize;
+
+        ShootRayFunctor
+        (
+            const point *_tgtPoints,
+            const point *_srcPoints,
+            const vector *_srcCf,
+            const faceData *_tgtFaces,
+            const faceData *_srcFaces,
+            label* _srcCount,
+            label* _tgtCount,
+            label* _srcAddr,
+            label* _tgtAddr,
+            const label _srcSize,
+            const label _tgtSize
+        ) :
+            tgtPoints(_tgtPoints),
+            srcPoints(_srcPoints),
+            srcCf(_srcCf),
+            tgtFaces(_tgtFaces),
+            srcFaces(_srcFaces),
+            rayFunctor(_tgtPoints, intersection::FULL_RAY, intersection::VECTOR),
+            normalFunctor(_srcPoints),
+            srcCount(_srcCount),
+            tgtCount(_tgtCount),
+            srcAddr(_srcAddr),
+            tgtAddr(_tgtAddr),
+            srcSize(_srcSize),
+            tgtSize(_tgtSize)
+        {}
+
+        template <typename Tuple>
+        __host__ __device__
+        void operator()(Tuple t)
+        {
+            const label tgtI = thrust::get<0>(t);
+            const label srcI = thrust::get<1>(t);
+
+            if (rayFunctor(tgtFaces[tgtI], srcCf[srcI], normalFunctor(srcFaces[srcI])).hit()) {
+                if (srcCount[srcI] < 15)
+                {
+                    srcAddr[srcSize*srcI + srcCount[srcI]] = tgtI;
```

Comment:
I think you are correct about this. However I see a problem. Imagine that a face contacts slightly with for example 15 faces and overlaps significantly with 16th face. In this case this algorithm will skip the most significant face in favour of less significant. You should check if you are ignoring face that is marginally overlapping.